apps/api/src/app/events/e2e/send-message-push.e2e.ts

```diff
@@ -0,0 +1,207 @@
+        _environmentId: session.environment._id,
+        channel: ChannelTypeEnum.PUSH,
+      });
+
+      expect(integrations.length).to.equal(2);
+    });
+
+    afterEach(async () => {
+      await executionDetailsRepository.delete({ _environmentId: session.environment._id });
+    });
+
+    it('should not create any message if subscriber has no configured channel', async () => {
+      await triggerEvent(session, template);
+
+      await session.awaitRunningJobs(template._id);
+
+      const messages = await messageRepository.find({
+        _environmentId: session.environment._id,
+        _templateId: template._id,
+        _subscriberId: session.subscriberId,
+      });
+
+      expect(messages.length).to.equal(0);
+
+      const executionDetails = await executionDetailsRepository.find({
+        _environmentId: session.environment._id,
+      });
+
+      expect(executionDetails.length).to.equal(5);
+      const noActiveChannel = executionDetails.find((ex) => ex.detail === DetailEnum.SUBSCRIBER_NO_ACTIVE_CHANNEL);
+      expect(noActiveChannel).to.be.ok;
+      expect(noActiveChannel?.providerId).to.equal('fcm');
+      const genericError = executionDetails.find((ex) => ex.detail === DetailEnum.NOTIFICATION_ERROR);
+      expect(genericError).to.be.ok;
+    });
+
+    it('should not create any message if subscriber has configured two providers without device tokens', async () => {
+      await updateCredentials(session, session.subscriberId, PushProviderIdEnum.FCM, []);
+      await updateCredentials(session, session.subscriberId, PushProviderIdEnum.EXPO, []);
+
+      await triggerEvent(session, template);
+
+      await session.awaitRunningJobs(template._id);
+
+      const messages = await messageRepository.find({
+        _environmentId: session.environment._id,
+        _templateId: template._id,
+        _subscriberId: session.subscriberId,
+      });
+
+      expect(messages.length).to.equal(0);
+
+      const executionDetails = await executionDetailsRepository.find({
+        _environmentId: session.environment._id,
+      });
+
+      expect(executionDetails.length).to.equal(6);
+      const fcm = executionDetails.find(
+        (ex) => ex.detail === DetailEnum.PUSH_MISSING_DEVICE_TOKENS && ex.providerId === PushProviderIdEnum.FCM
+      );
+      expect(fcm).to.be.ok;
+      const expo = executionDetails.find(
+        (ex) => ex.detail === DetailEnum.PUSH_MISSING_DEVICE_TOKENS && ex.providerId === PushProviderIdEnum.EXPO
+      );
+      expect(expo).to.be.ok;
+      const genericError = executionDetails.find((ex) => ex.detail === DetailEnum.NOTIFICATION_ERROR);
+      expect(genericError).to.be.ok;
+    });
+
+    it('should not create any message if subscriber has configured one provider without device tokens and the other has wrong credentials', async () => {
+      await updateCredentials(session, session.subscriberId, PushProviderIdEnum.FCM, ['deviceToken']);
```

Comment:
how come this one is an wrong credential?

Reply:
Because the it is not a valid device token. I will rename it to avoid confusion.